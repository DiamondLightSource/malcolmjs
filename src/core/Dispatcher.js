/*
 * React.js Starter Kit
 * Copyright (c) 2014 Konstantin Tarkus (@koistya), KriaSoft LLC.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import Flux from 'flux';
import PayloadSources from '../constants/PayloadSources';
import assign from 'react/lib/Object.assign';

/**
 * A singleton that operates as the central hub for application updates.
 * For more information visit https://facebook.github.io/flux/
 */
let Dispatcher = assign(new Flux.Dispatcher(), {

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the server.
   */
  handleServerAction(action) {
    var payload = {
      source: PayloadSources.SERVER_ACTION,
      action: action
    };
    this.dispatch(payload);
  },

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the view.
   */
  handleViewAction(action) {
    var payload = {
      source: PayloadSources.VIEW_ACTION,
      action: action
    };
    this.dispatch(payload);
  }

});

export default Dispatcher;
