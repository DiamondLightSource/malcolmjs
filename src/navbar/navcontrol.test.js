import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import NavControl, { NavSelector } from './navcontrol.component';

describe('NavControl', () => {
  let shallow;
  let mount;

  let nav;

  const mockTheme = { palette: { text: { primary: '#fff' } } };

  beforeEach(() => {
    shallow = createShallow();
    mount = createMount();

    nav = {
      path: 'PANDA:mri',
      children: { layout: { label: 'layout' }, table: { label: 'table' } },
      label: 'PANDA',
      parent: {
        basePath: '/',
        children: { 'PANDA:mri': { label: 'PANDA' } },
      },
    };
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('renders correctly', () => {
    const wrapper = shallow(
      <NavControl nav={nav} navigateToChild={() => {}} />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders correctly for final', () => {
    const wrapper = shallow(
      <NavControl nav={nav} navigateToChild={() => {}} isFinalNav />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('selector component renders correctly', () => {
    const wrapper = shallow(
      <NavSelector
        handleClick={() => {}}
        childElements={nav.children}
        childElementLabels={nav.childrenLabels}
        anchorEl="âš“"
        handleClose={() => {}}
        navigateToChild={() => {}}
        icon={<div />}
        theme={mockTheme}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('disables menu item if children is empty', () => {
    nav.parent.children = [];

    const wrapper = shallow(
      <NavSelector
        handleClick={() => {}}
        childElements={[]}
        childElementLabels={[]}
        anchorEl=""
        handleClose={() => {}}
        navigateToChild={() => {}}
        icon={<div />}
        theme={mockTheme}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('updates the route if the link is clicked', () => {
    const navigateFunction = jest.fn();

    const wrapper = mount(
      <NavControl nav={nav} navigateToChild={navigateFunction} />
    );

    wrapper.find('Typography').simulate('click');

    expect(navigateFunction.mock.calls.length).toEqual(1);
  });

  it('handles the nav drop down being clicked', () => {
    const wrapper = mount(<NavControl nav={nav} navigateToChild={() => {}} />);
    wrapper.find('IconButton').simulate('click');

    expect(
      wrapper
        .find(NavControl)
        .childAt(0)
        .instance().state.anchorEl
    ).toBeInstanceOf(HTMLButtonElement);
  });

  it('clicking on a menu item should close the menu and navigate to child', () => {
    const navigateFunction = jest.fn();
    const wrapper = mount(
      <NavControl nav={nav} navigateToChild={navigateFunction} />
    );

    // open the menu
    wrapper.find('IconButton').simulate('click');
    expect(
      wrapper
        .find(NavControl)
        .childAt(0)
        .instance().state.anchorEl
    ).toBeInstanceOf(HTMLButtonElement);

    wrapper
      .find('MenuItem')
      .at(0)
      .simulate('click');
    expect(
      wrapper
        .find(NavControl)
        .childAt(0)
        .instance().state.anchorEl
    ).toBeNull();
    expect(navigateFunction.mock.calls.length).toEqual(1);
  });

  it('renders correctly when nav element is errored but has valid siblings', () => {
    nav.path = 'NotPANDA';
    const wrapper = shallow(
      <NavControl nav={nav} navigateToChild={() => {}} />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders correctly when nav element is errored and has no siblings', () => {
    nav.path = 'NotPANDA';
    nav.parent.children = undefined;
    const wrapper = shallow(
      <NavControl nav={nav} navigateToChild={() => {}} />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
