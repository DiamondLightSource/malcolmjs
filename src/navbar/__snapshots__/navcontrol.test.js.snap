// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavControl disables menu item if children is empty 1`] = `
<div
  style={
    Object {
      "paddingRight": "12px",
    }
  }
>
  <WithStyles(IconButton)
    data-cy="navmenu"
    disabled={true}
    onClick={[Function]}
    style={
      Object {
        "height": "32px",
        "width": "32px",
      }
    }
  >
    <div />
  </WithStyles(IconButton)>
  <WithStyles(Menu)
    anchorEl=""
    id="simple-menu"
    onClose={[Function]}
    open={false}
  />
</div>
`;

exports[`NavControl renders correctly 1`] = `
<div
  className="NavControl-container-1"
>
  <WithStyles(Typography)
    className="NavControl-currentLink-2"
    onClick={[Function]}
    variant="subheading"
  >
    PANDA
  </WithStyles(Typography)>
  <Component
    anchorEl={null}
    childElementLabels={
      Array [
        "PANDA",
      ]
    }
    childElements={
      Array [
        "PANDA:mri",
      ]
    }
    handleClick={[Function]}
    handleClose={[Function]}
    icon={<pure(KeyboardArrowDown) />}
    navigateToChild={[Function]}
  />
</div>
`;

exports[`NavControl renders correctly for final 1`] = `
<div
  className="NavControl-container-1"
>
  <Component
    anchorEl={null}
    childElementLabels={
      Array [
        "layout",
        "table",
      ]
    }
    childElements={
      Array [
        "layout",
        "table",
      ]
    }
    handleClick={[Function]}
    handleClose={[Function]}
    icon={<pure(MoreHoriz) />}
    navigateToChild={[Function]}
  />
</div>
`;

exports[`NavControl selector component renders correctly 1`] = `
<div
  style={
    Object {
      "paddingRight": "12px",
    }
  }
>
  <WithStyles(IconButton)
    data-cy="navmenu"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "height": "32px",
        "width": "32px",
      }
    }
  >
    <div />
  </WithStyles(IconButton)>
  <WithStyles(Menu)
    anchorEl="âš“"
    id="simple-menu"
    onClose={[Function]}
    open={true}
  >
    <WithStyles(MenuItem)
      key="layout"
      onClick={[Function]}
    >
      layout
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      key="table"
      onClick={[Function]}
    >
      table
    </WithStyles(MenuItem)>
  </WithStyles(Menu)>
</div>
`;
